File: modules/dock-ble/android/src/main/java/expo/modules/dockble/DockBleModule.kt
package expo.modules.dockble

import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * DockBleModule (Android) — Phase-3 BLE dock integration skeleton.
 *
 * CURRENT STATE: All BLE methods return stub/simulated values.
 * The real state-machine logic lives in TypeScript (DockStateMachine.ts +
 * DockService.ts). This native module handles:
 *   1. BLE permission management (BLUETOOTH_SCAN / BLUETOOTH_CONNECT).
 *   2. Android BLE scan / connect skeletons via BluetoothLeScanner / BluetoothGatt.
 *   3. EventEmitter surface for forwarding dock BLE events to JS.
 *
 * BLE PLUG-IN INSTRUCTIONS
 * ─────────────────────────
 * When real BLE hardware is available, fill in the MARK: BLE sections below:
 *
 * 1. Inject BluetoothAdapter via appContext.reactContext:
 *      val bluetoothManager = context.getSystemService(BluetoothManager::class.java)
 *      val bluetoothAdapter = bluetoothManager?.adapter
 *
 * 2. In startScan(): call
 *      bluetoothAdapter.bluetoothLeScanner.startScan(
 *        listOf(ScanFilter.Builder().setServiceUuid(ParcelUuid(DOCK_SERVICE_UUID)).build()),
 *        ScanSettings.Builder().build(),
 *        scanCallback
 *      )
 *    Collect ScanResult.device in a map, stop after timeoutMs, resolve.
 *
 * 3. In connectPeripheral(peripheralId): retrieve the BluetoothDevice by address
 *    and call device.connectGatt(context, false, gattCallback).
 *    In onConnectionStateChange(GATT_SUCCESS, STATE_CONNECTED): discover services.
 *    In onServicesDiscovered(): resolve the connect promise.
 *
 * 4. In subscribeToFrameEvents(peripheralId): enable notifications on the
 *    frame-alignment and dock-status characteristics:
 *      gatt.setCharacteristicNotification(characteristic, true)
 *      val descriptor = characteristic.getDescriptor(CCCD_UUID)
 *      descriptor.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
 *      gatt.writeDescriptor(descriptor)
 *    In onCharacteristicChanged: decode the byte payload and call
 *      sendEvent("dockEvent", mapOf("type" to "frameAligned", ...))
 *
 * 5. In readBatteryLevel(peripheralId): gatt.readCharacteristic(batteryChar).
 *    Resolve in onCharacteristicRead with value[0].toInt() & 0xFF.
 *
 * Known dock BLE service UUID (to be assigned by firmware team):
 *   val DOCK_SERVICE_UUID = UUID.fromString("XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX")
 * Known frame-alignment characteristic UUID:
 *   val FRAME_ALIGNMENT_CHAR_UUID = UUID.fromString("YYYYYYYY-YYYY-YYYY-YYYY-YYYYYYYYYYYY")
 * Known CCCD descriptor UUID (standard):
 *   val CCCD_UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
 *
 * Required AndroidManifest permissions (add to app's AndroidManifest.xml):
 *   <!-- API 31+ -->
 *   <uses-permission android:name="android.permission.BLUETOOTH_SCAN"
 *     android:usesPermissionFlags="neverForLocation" />
 *   <uses-permission android:name="android.permission.BLUETOOTH_CONNECT" />
 *   <!-- Legacy API 28-30 -->
 *   <uses-permission android:name="android.permission.BLUETOOTH" android:maxSdkVersion="30" />
 *   <uses-permission android:name="android.permission.BLUETOOTH_ADMIN" android:maxSdkVersion="30" />
 *   <uses-feature android:name="android.hardware.bluetooth_le" android:required="false" />
 */
class DockBleModule : Module() {

    // BLE PLUG-IN: private var bluetoothAdapter: BluetoothAdapter? = null
    // BLE PLUG-IN: private var bluetoothGatt: BluetoothGatt? = null
    // BLE PLUG-IN: private val discoveredDevices: MutableMap<String, BluetoothDevice> = mutableMapOf()

    override fun definition() = ModuleDefinition {
        Name("DockBle")

        // MARK: - EventEmitter
        //
        // "dockEvent" is fired by the native layer when a BLE GATT characteristic
        // notification arrives from the dock firmware.
        //
        // Payload shape (mirrors DockEvent + DockEventPayload in DockBle.types.ts):
        //   {
        //     "type": "frameAligned" | "jamDetected" | "lowBattery" | "disconnected",
        //     "timestamp": "<ISO 8601>",
        //     "payload": {
        //       "frameIndex": Int?,
        //       "batteryLevel": Int?,
        //       "reason": String?,
        //       "rssi": Int?
        //     }
        //   }
        //
        // BLE PLUG-IN: Call `sendEvent("dockEvent", mapOf(...))` inside the
        //   BluetoothGattCallback.onCharacteristicChanged callback after decoding
        //   the characteristic notification bytes.
        Events("dockEvent")

        // MARK: - requestBlePermissions() -> String

        // Returns "granted" | "denied" | "restricted".
        // STUB: returns "granted" unconditionally (simulation mode does not need BLE).
        //
        // BLE PLUG-IN: Check / request BLUETOOTH_SCAN + BLUETOOTH_CONNECT at runtime:
        //   val context = appContext.reactContext ?: return "restricted"
        //   val scanGranted = ContextCompat.checkSelfPermission(context,
        //     Manifest.permission.BLUETOOTH_SCAN) == PackageManager.PERMISSION_GRANTED
        //   val connectGranted = ContextCompat.checkSelfPermission(context,
        //     Manifest.permission.BLUETOOTH_CONNECT) == PackageManager.PERMISSION_GRANTED
        //   return if (scanGranted && connectGranted) "granted" else "denied"
        AsyncFunction("requestBlePermissions") Coroutine { ->
            "granted"
        }

        // MARK: - startScan(timeoutMs) -> List<String>

        // STUB: resolves immediately with an empty list.
        //
        // BLE PLUG-IN: Start BluetoothLeScanner.startScan() with a ScanFilter on the
        //   dock service UUID. Collect discovered device addresses in discoveredDevices.
        //   Stop the scan after timeoutMs and return the collected address strings.
        AsyncFunction("startScan") Coroutine { _: Int ->
            // BLE PLUG-IN: Start BLE scan here.
            emptyList<String>()
        }

        // MARK: - stopScan() -> Unit

        // STUB: no-op.
        //
        // BLE PLUG-IN: bluetoothAdapter?.bluetoothLeScanner?.stopScan(scanCallback)
        AsyncFunction("stopScan") Coroutine { ->
            // BLE PLUG-IN: stop the BLE scan.
        }

        // MARK: - connectPeripheral(peripheralId) -> String

        // STUB: returns "connected" immediately.
        //
        // BLE PLUG-IN:
        //   val device = discoveredDevices[peripheralId]
        //     ?: throw PeripheralNotFoundException()
        //   bluetoothGatt = device.connectGatt(context, false, gattCallback)
        //   // Resolve in gattCallback.onConnectionStateChange(STATE_CONNECTED)
        AsyncFunction("connectPeripheral") Coroutine { _: String ->
            "connected"
        }

        // MARK: - disconnectPeripheral(peripheralId) -> Unit

        // STUB: no-op.
        //
        // BLE PLUG-IN:
        //   bluetoothGatt?.disconnect()
        //   bluetoothGatt?.close()
        //   bluetoothGatt = null
        AsyncFunction("disconnectPeripheral") Coroutine { _: String ->
            // BLE PLUG-IN: disconnect and close GATT.
        }

        // MARK: - readBatteryLevel(peripheralId) -> Int

        // STUB: returns simulated battery level 85.
        //
        // BLE PLUG-IN:
        //   val service = bluetoothGatt?.getService(BATTERY_SERVICE_UUID)
        //   val characteristic = service?.getCharacteristic(BATTERY_LEVEL_CHAR_UUID)
        //   bluetoothGatt?.readCharacteristic(characteristic)
        //   // Resolve in gattCallback.onCharacteristicRead with value[0].toInt() and 0xFF
        AsyncFunction("readBatteryLevel") Coroutine { _: String ->
            85 // Simulated battery level
        }

        // MARK: - subscribeToFrameEvents(peripheralId) -> Unit

        // STUB: no-op. In simulation mode, events come from SimulatedDock via JS timers.
        //
        // BLE PLUG-IN:
        //   val frameChar = bluetoothGatt
        //     ?.getService(DOCK_SERVICE_UUID)
        //     ?.getCharacteristic(FRAME_ALIGNMENT_CHAR_UUID)
        //     ?: throw PeripheralNotFoundException()
        //   bluetoothGatt?.setCharacteristicNotification(frameChar, true)
        //   val descriptor = frameChar.getDescriptor(CCCD_UUID)
        //   descriptor.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
        //   bluetoothGatt?.writeDescriptor(descriptor)
        //   // In onCharacteristicChanged: decode + sendEvent("dockEvent", payload)
        AsyncFunction("subscribeToFrameEvents") Coroutine { _: String ->
            // BLE PLUG-IN: enable GATT characteristic notifications here.
        }
    }
}

// MARK: - Exceptions

internal class BlePermissionDeniedException :
    CodedException(message = "Bluetooth permission denied")

internal class BleUnavailableException :
    CodedException(message = "Bluetooth is not available on this device")

internal class PeripheralNotFoundException :
    CodedException(message = "Dock peripheral not found or not connected")

File: modules/dock-ble/ios/DockBleModule.swift
//
//  DockBleModule.swift
//  dock-ble (local Expo module)
//
//  Phase-3 BLE dock integration skeleton — iOS (CoreBluetooth).
//
//  CURRENT STATE: All BLE methods return stub/simulated values.
//  The real state-machine logic lives in TypeScript (DockStateMachine.ts +
//  DockService.ts). This native module handles:
//    1. BLE permission management (CBCentralManager authorization).
//    2. CoreBluetooth scan / connect skeletons.
//    3. EventEmitter surface for forwarding dock BLE events to JS.
//
//  BLE PLUG-IN INSTRUCTIONS
//  ─────────────────────────
//  When real BLE hardware is available, fill in the MARK: BLE sections below:
//
//  1. Import CoreBluetooth and implement CBCentralManagerDelegate +
//     CBPeripheralDelegate.
//
//  2. In startScan(): call
//       centralManager.scanForPeripherals(
//         withServices: [CBUUID(string: DOCK_SERVICE_UUID)],
//         options: nil
//       )
//     and collect discovered peripherals in a dict; resolve the promise
//     in centralManagerDidDiscover peripheral callback.
//
//  3. In connectPeripheral(peripheralId:): retrieve the peripheral from the
//     dict and call centralManager.connect(peripheral, options: nil).
//     In centralManager(_:didConnect:): resolve the promise and call
//     subscribeToFrameEvents if auto-subscribe is desired.
//
//  4. In subscribeToFrameEvents(peripheralId:): after GATT service discovery,
//     call peripheral.setNotifyValue(true, for: frameAlignmentCharacteristic).
//     In peripheral(_:didUpdateValueFor:): decode the byte payload and call
//     sendEvent("dockEvent", body: ["type": "frameAligned", ...]).
//
//  5. In readBatteryLevel(peripheralId:): read the Battery Service
//     characteristic (UUID 0x2A19) and resolve with the UInt8 value.
//
//  Known dock BLE service UUID (to be assigned by firmware team):
//    let DOCK_SERVICE_UUID = "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX"
//  Known frame-alignment characteristic UUID:
//    let FRAME_ALIGNMENT_CHAR_UUID = "YYYYYYYY-YYYY-YYYY-YYYY-YYYYYYYYYYYY"
//  Known dock-status (jam/battery) characteristic UUID:
//    let DOCK_STATUS_CHAR_UUID = "ZZZZZZZZ-ZZZZ-ZZZZ-ZZZZ-ZZZZZZZZZZZZ"
//

import ExpoModulesCore
// BLE PLUG-IN: import CoreBluetooth

// MARK: - Exceptions

internal final class BlePermissionDeniedException: Exception {
    override var reason: String { "Bluetooth permission denied" }
}

internal final class BleUnavailableException: Exception {
    override var reason: String { "Bluetooth is not available on this device" }
}

internal final class PeripheralNotFoundException: Exception {
    override var reason: String { "Dock peripheral not found or not connected" }
}

// MARK: - Module

public class DockBleModule: Module {

    // BLE PLUG-IN: private var centralManager: CBCentralManager?
    // BLE PLUG-IN: private var discoveredPeripherals: [String: CBPeripheral] = [:]
    // BLE PLUG-IN: private var connectedPeripheral: CBPeripheral?
    // BLE PLUG-IN: private var scanCompletionHandlers: [(Result<[String], Error>) -> Void] = []

    public func definition() -> ModuleDefinition {
        Name("DockBle")

        // MARK: - EventEmitter
        //
        // "dockEvent" is fired by the native layer when a BLE characteristic
        // notification arrives from the dock firmware.
        //
        // Payload shape (mirrors DockEvent + DockEventPayload in DockBle.types.ts):
        //   {
        //     "type": "frameAligned" | "jamDetected" | "lowBattery" | "disconnected",
        //     "timestamp": "<ISO 8601>",
        //     "payload": {
        //       "frameIndex": Int?,
        //       "batteryLevel": Int?,
        //       "reason": String?,
        //       "rssi": Int?
        //     }
        //   }
        //
        // BLE PLUG-IN: Call `sendEvent("dockEvent", body: payload)` inside the
        //   CBPeripheralDelegate.peripheral(_:didUpdateValueFor:) callback after
        //   decoding the characteristic notification bytes.
        Events("dockEvent")

        // MARK: - requestBlePermissions() -> String

        // Returns "granted" | "denied" | "restricted".
        // STUB: returns "granted" unconditionally (simulation mode does not need BLE).
        //
        // BLE PLUG-IN: Replace with real CBCentralManager authorization check:
        //   switch CBCentralManager.authorization {
        //   case .allowedAlways: return "granted"
        //   case .denied: return "denied"
        //   case .restricted: return "restricted"
        //   default: // .notDetermined — prompt via CBCentralManager init
        //   }
        AsyncFunction("requestBlePermissions") { () -> String in
            return "granted"
        }

        // MARK: - startScan(timeoutMs) -> [String]

        // Scans for dock peripherals and returns discovered peripheral ID strings.
        // STUB: resolves immediately with an empty array.
        //
        // BLE PLUG-IN: Start CBCentralManager.scanForPeripherals(withServices:).
        //   Collect results in centralManagerDidDiscover. Stop after timeoutMs and
        //   resolve with the collected peripheral UUID strings.
        AsyncFunction("startScan") { (timeoutMs: Int) -> [String] in
            // BLE PLUG-IN: Start BLE scan here.
            // Return empty array in stub — simulation mode bypasses this.
            return []
        }

        // MARK: - stopScan() -> Void

        // STUB: no-op.
        //
        // BLE PLUG-IN: centralManager.stopScan()
        AsyncFunction("stopScan") { () -> Void in
            // BLE PLUG-IN: centralManager?.stopScan()
        }

        // MARK: - connectPeripheral(peripheralId) -> String

        // STUB: returns "connected" immediately (simulation mode uses SimulatedDock
        // and does not call through to native BLE).
        //
        // BLE PLUG-IN:
        //   guard let peripheral = discoveredPeripherals[peripheralId] else {
        //     throw PeripheralNotFoundException()
        //   }
        //   centralManager.connect(peripheral, options: nil)
        //   // Resolve in centralManager(_:didConnect:) callback.
        AsyncFunction("connectPeripheral") { (peripheralId: String) -> String in
            return "connected"
        }

        // MARK: - disconnectPeripheral(peripheralId) -> Void

        // STUB: no-op.
        //
        // BLE PLUG-IN:
        //   if let peripheral = connectedPeripheral {
        //     centralManager.cancelPeripheralConnection(peripheral)
        //   }
        AsyncFunction("disconnectPeripheral") { (peripheralId: String) -> Void in
            // BLE PLUG-IN: centralManager?.cancelPeripheralConnection(connectedPeripheral)
        }

        // MARK: - readBatteryLevel(peripheralId) -> Int

        // STUB: returns simulated battery level 85.
        //
        // BLE PLUG-IN:
        //   peripheral.readValue(for: batteryLevelCharacteristic)
        //   // Resolve in peripheral(_:didUpdateValueFor:) for the Battery Service char.
        //   // Return value[0] as Int (0–100).
        AsyncFunction("readBatteryLevel") { (peripheralId: String) -> Int in
            return 85 // Simulated battery level
        }

        // MARK: - subscribeToFrameEvents(peripheralId) -> Void

        // STUB: no-op. In simulation mode, events come from SimulatedDock via JS timers.
        //
        // BLE PLUG-IN:
        //   peripheral.setNotifyValue(true, for: frameAlignmentCharacteristic)
        //   peripheral.setNotifyValue(true, for: dockStatusCharacteristic)
        //   // In peripheral(_:didUpdateValueFor:), decode payload bytes and call:
        //   //   sendEvent("dockEvent", body: decodedPayload)
        //   // where decodedPayload matches the DockEvent shape above.
        AsyncFunction("subscribeToFrameEvents") { (peripheralId: String) -> Void in
            // BLE PLUG-IN: enable characteristic notifications here.
        }
    }
}
